pkg/mimirtool/client/backfill.go

```diff
@@ -0,0 +1,196 @@
+	for _, b := range blocks {
+		if err := c.backfillBlock(ctx, b, logger); err != nil {
+			if errors.Is(err, errConflict) {
+				level.Warn(logger).Log("msg", "failed uploading block since it already exists on server",
+					"path", b)
+				alreadyExists = append(alreadyExists, b)
+			} else {
+				level.Warn(logger).Log("msg", "failed uploading block", "path", b, "err", err)
+				failed = append(failed, b)
+			}
+			continue
+		}
+
+		level.Info(logger).Log("msg", "successfully uploaded block", "path", b)
+		succeeded = append(succeeded, b)
+	}
+
+	level.Info(logger).Log("msg", "finished uploading block(s)", "succeeded", len(succeeded),
+		"already_exists", len(alreadyExists), "failed", len(failed))
+
+	if len(failed) > 0 {
+		return fmt.Errorf("failed to upload %d block(s)", len(failed))
+	}
+
+	return nil
+}
+
+func closeResp(resp *http.Response) {
+	// Drain and close the body to let the transport reuse the connection
+	io.Copy(io.Discard, resp.Body)
+	resp.Body.Close()
+}
+
+func (c *MimirClient) backfillBlock(ctx context.Context, blockDir string, logger log.Logger) error {
+	blockMeta, err := getBlockMeta(blockDir)
+	if err != nil {
+		return err
+	}
+
+	blockID := blockMeta.ULID.String()
+	logger = log.With(logger, "user", c.id, "block", blockID)
+
+	level.Info(logger).Log("msg", "making request to start block upload")
+
+	blockPrefix := path.Join("/api/v1/upload/block", url.PathEscape(blockID))
+
+	buf := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buf).Encode(blockMeta); err != nil {
+		return errors.Wrap(err, "failed to JSON encode payload")
+	}
+	resp, err := c.doRequest(blockPrefix, http.MethodPost, buf, int64(buf.Len()))
+	if err != nil {
+		return errors.Wrap(err, "request to start block upload failed")
+	}
+	closeResp(resp)
+	if resp.StatusCode/100 != 2 {
+		return fmt.Errorf("request to start block upload failed, with HTTP status %d %s",
+			resp.StatusCode, resp.Status)
+	}
+
+	// Upload each block file
+	for _, tf := range blockMeta.Thanos.Files {
+		if tf.RelPath == block.MetaFilename {
+			// Don't upload meta file in this step
+			continue
+		}
+
+		pth := filepath.Join(blockDir, filepath.FromSlash(tf.RelPath))
+		f, err := os.Open(pth)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open %q", pth)
+		}
+		defer func() {
+			_ = f.Close()
+		}()
+
+		st, err := f.Stat()
+		if err != nil {
+			return errors.Wrapf(err, "failed to get file info for %q", pth)
+		}
+
+		escapedPath := url.QueryEscape(tf.RelPath)
+		level.Info(logger).Log("msg", "uploading block file", "path", pth, "size", st.Size())
+		resp, err := c.doRequest(path.Join(blockPrefix, fmt.Sprintf("files?path=%s", escapedPath)),
+			http.MethodPost, f, st.Size())
+		if err != nil {
+			return errors.Wrapf(err, "request to upload file %q failed", pth)
+		}
+		closeResp(resp)
+		if resp.StatusCode/100 != 2 {
+			return fmt.Errorf("request to upload block file failed, with HTTP status %d %s",
+				resp.StatusCode, resp.Status)
+		}
```

Comment:
Perhaps we should extract single file upload to a function, so that deferred `f.Close()` doesn't wait until all files are uploaded. Otherwise we could run out of file descriptors (eg. with low open files limit, block with lot of chunks files, or when we do parallel uploads of blocks).

Reply:
Good point @pstibrany - done.